src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,220 @@
+import org.xml.sax.SAXException;
+
+/**
+ * <p>This listener creates an additional report for pioneer extension and validates it against the "pioneerreport.xsd".
+ * This is done because the default junit test report can't / should not be extended.
+ * A dynamic test report which shall be used by all test tools is planed, but not yet realized, see
+ * <a href="https://github.com/ota4j-team/opentest4j/issues/9">issue #9 at opentest4j</a> for current progress.</p>
+ *
+ * <p>The report contains test containers and its test methods. For each object properties, based on published
+ * report entries are reported. This allows extension to report additional values in the report.
+ * for test cases the result of the test is (of course) reported.</p>
+ *
+ * @see <a href="https://junit.org/junit5/docs/current/api/org.junit.platform.launcher/org/junit/platform/launcher/TestPlan.html">JUnit Docs</a>
+ */
+public class PioneerReportListener implements TestExecutionListener {
+
+	TestPlan storedTestPlan;
+	PioneerReport pioneerReport;
+
+	ConcurrentHashMap<String, List<Property>> propertyCache = new ConcurrentHashMap<>();
+	ConcurrentHashMap<String, String> testStatusCache = new ConcurrentHashMap<>();
+
+	@Override
+	public void reportingEntryPublished(TestIdentifier testIdentifier, ReportEntry entry) {
+		String id = testIdentifier.getUniqueId();
+
+		// Read value from published entry
+		for (Map.Entry<String, String> singleEntry : entry.getKeyValuePairs().entrySet()) {
+			Property property = new Property(singleEntry.getKey(), singleEntry.getValue());
+			// Add to cache
+			if (propertyCache.containsKey(id)) {
+				propertyCache.get(id).add(property);
+			} else {
+				List<Property> currentProperties = new ArrayList<>();
+				currentProperties.add(property);
+				propertyCache.put(id, currentProperties);
+			}
+		}
+	}
+
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		if (testIdentifier.isTest()) {
+			// Store result in cache
+			testStatusCache.put(testIdentifier.getUniqueId(), testExecutionResult.getStatus().toString());
+		}
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan testPlan) {
+		// Store plan to access it from methods
+		this.storedTestPlan = testPlan;
+		pioneerReport = new PioneerReport();
+
+		Set<TestIdentifier> allRoots = testPlan.getRoots();
+
+		for (TestIdentifier identifier : allRoots) {
+			if (identifier.isContainer()) {
+				pioneerReport.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				pioneerReport.getTestcase().add(processTest(identifier));
+			}
+		}
+		writeReport(pioneerReport);
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcontainer} for the report.
+	 *
+	 * @param container Object to be converted into {@code Testcontainer}
+	 * @return Created object
+	 */
+	Testcontainer processContainer(TestIdentifier container) {
```

Comment:
Shouldn't this method be `private`?

Reply:
I don't create private methods, so tests are able to access them.